File: pdk/commands/doctor.py
"""`pdk doctor` — check node version, environment health, and chain liveness."""

from __future__ import annotations

import time

import typer
from rich.console import Console
from rich.markup import escape
from rich.table import Table

from pdk.config import DEFAULT_NODE_URL, EXPLORER_URL
from pdk.core.chain import connect
from pdk.core.decoder import strip_control_chars

console = Console()


def run(
    node: str = typer.Option(DEFAULT_NODE_URL, "--node", help="Portaldot node WS endpoint."),
    liveness: bool = typer.Option(
        True, "--liveness/--no-liveness", help="Check that the chain is producing blocks (waits ~7s)."
    ),
) -> None:
    """Report node version, runtime info, ink! compatibility, and chain liveness."""
    try:
        substrate = connect(node)
        substrate.init_runtime()  # load runtime + metadata (lazy by default)
    except Exception as exc:  # noqa: BLE001 — surface any connection failure plainly
        console.print(f"[red]Cannot reach a Portaldot node at {node}[/red]")
        console.print(f"[dim]{exc}[/dim]")
        console.print("Start a node with [bold]pdk up[/bold] (run the node in WSL on Windows; pdk itself runs natively).")
        raise typer.Exit(code=1)

    pallet_names = [pallet.name for pallet in substrate.metadata.pallets]
    table = Table(title="pdk doctor — Portaldot node health", show_header=False)
    table.add_row("Endpoint", node)
    # `substrate.chain` is the node's self-reported display name via the
    # system_chain RPC — free text set in whatever chain spec the node
    # operator configured, with no identifier-syntax constraint at all
    # (unlike a pallet/error name). A malicious/misconfigured node needs
    # no transaction, no compromise — just its own RPC response — to put
    # markup or a raw terminal escape sequence in front of every `pdk
    # doctor --node <that-endpoint>` user.
    table.add_row("Chain", escape(strip_control_chars(str(substrate.chain or "unknown"))))
    table.add_row("Runtime version", escape(strip_control_chars(str(substrate.runtime_version))))
    table.add_row("Pallets", str(len(pallet_names)))

    has_contracts = "Contracts" in pallet_names
    if has_contracts:
        table.add_row(
            "Contracts pallet",
            "present — node v2.0.0 ships contracts API v5, which supports "
            "[bold]ink! 3.x only[/bold] (ink! 4.x+ will not deploy)",
        )
    else:
        table.add_row("Contracts pallet", "not found — use native pallets, not ink!")

    table.add_row("Explorer", EXPLORER_URL)
    # AI status row - lets users see at a glance whether AI auto-on will trigger
    # on debug/explain/simulate/report. Computed locally; never hits the network.
    from pdk.core.ai import ai_status
    ai = ai_status()
    if ai["configured"]:
        table.add_row("AI (optional)",
                      f"[green]configured[/green] - {ai['key_preview']} via {ai['model']} "
                      f"(auto-runs on debug/explain/simulate/report)")
    else:
        table.add_row("AI (optional)",
                      "not configured - run [bold]pdk ai-setup[/bold] to enable")
    console.print(table)

    if liveness:
        first = substrate.get_block_number(substrate.get_chain_head())
        time.sleep(7)  # longer than one 6s block, so a healthy chain advances
        second = substrate.get_block_number(substrate.get_chain_head())
        if second > first:
            console.print(f"[green]✓ chain is producing blocks[/green] (#{first} → #{second})")
        else:
            console.print(f"[red]⚠ chain is stalled at #{first} — no new blocks are being produced.[/red]")
            console.print("[dim]A dev chain DB can wedge (BABE 'Unexpected epoch change'). Reset it:[/dim]")
            console.print("[dim]  portaldot_dev purge-chain --dev -y   # then restart the node[/dim]")
            raise typer.Exit(code=1)

File: pdk-ts/src/commands/doctor.ts
/**
 * `pdk-ts doctor` — health probe against a Portaldot (or any Substrate) node.
 *
 * Ports the shape of Python `pdk doctor` to TypeScript. Fetches:
 *   - endpoint / chain name / runtime version
 *   - pallet count (from metadata)
 *   - contracts pallet presence (ink! compat signal)
 *   - liveness: is the chain actually producing blocks, not just reachable
 *     (mirrors Python's `--liveness/--no-liveness`, default on)
 *
 * Exit code 0 on success, 1 if the node is unreachable, the metadata
 * fetch fails, or the chain is stalled (liveness check fails).
 */

import pc from 'picocolors';
import {getApi, closeApi} from '../core/chain.js';
import {resolveNode} from '../core/config.js';
import {humanizeChainError, stripControlChars} from '../core/errors.js';

export interface DoctorOptions {
  node?: string;
  json?: boolean;
  timeout?: string; // seconds, from commander
  liveness?: boolean; // default true — commander maps --no-liveness to false
}

export interface DoctorReport {
  endpoint: string;
  chain: string;
  nodeName: string;
  nodeVersion: string;
  specName: string;
  specVersion: number;
  palletCount: number;
  contractsPresent: boolean;
  liveness?: {checked: boolean; producing: boolean; blockBefore: number; blockAfter: number};
}

// A healthy chain advances at least once within one block-time; wait
// longer than a single ~6s block so we don't false-positive "stalled"
// on a chain that just happened to sample between blocks.
const LIVENESS_WAIT_MS = 7_000;

export async function collectReport(
  node: string,
  timeoutMs?: number,
  checkLiveness = true,
): Promise<DoctorReport> {
  const api = await getApi(node, timeoutMs);
  const [chain, nodeName, nodeVersion] = await Promise.all([
    api.rpc.system.chain(),
    api.rpc.system.name(),
    api.rpc.system.version(),
  ]);
  const runtimeVersion = api.runtimeVersion;
  const palletCount = api.runtimeMetadata.asLatest.pallets.length;
  const contractsPresent = api.runtimeMetadata.asLatest.pallets.some(
    (p) => p.name.toString().toLowerCase() === 'contracts',
  );

  let liveness: DoctorReport['liveness'];
  if (checkLiveness) {
    const before = (await api.rpc.chain.getHeader()).number.toNumber();
    await new Promise((resolve) => setTimeout(resolve, LIVENESS_WAIT_MS));
    const after = (await api.rpc.chain.getHeader()).number.toNumber();
    liveness = {checked: true, producing: after > before, blockBefore: before, blockAfter: after};
  }

  return {
    endpoint: node,
    chain: chain.toString(),
    nodeName: nodeName.toString(),
    nodeVersion: nodeVersion.toString(),
    specName: runtimeVersion.specName.toString(),
    specVersion: runtimeVersion.specVersion.toNumber(),
    palletCount,
    contractsPresent,
    liveness,
  };
}

export function renderText(r: DoctorReport): string {
  // r.chain/nodeName/nodeVersion/specName are the node's own
  // self-reported RPC strings (system_chain, system_name,
  // system_version, runtime spec name) — free text with no identifier-
  // syntax constraint. A malicious/misconfigured node needs no
  // transaction, no compromise, just its own RPC response to embed a
  // raw terminal escape sequence in front of every user who points
  // `pdk-ts doctor` at it. picocolors doesn't parse markup (unlike
  // Rich) but does not filter control bytes either — strip them here.
  const rows: [string, string][] = [
    ['Endpoint', r.endpoint],
    ['Chain', stripControlChars(r.chain)],
    ['Node', stripControlChars(`${r.nodeName} ${r.nodeVersion}`)],
    ['Runtime', stripControlChars(`${r.specName} ${r.specVersion}`)],
    ['Pallets', String(r.palletCount)],
    [
      'Contracts pallet',
      r.contractsPresent
        ? pc.green('present') + pc.dim(' — ink! contracts supported')
        : pc.yellow('absent'),
    ],
  ];
  const width = Math.max(...rows.map(([k]) => k.length));
  const base = rows.map(([k, v]) => `  ${pc.dim(k.padEnd(width))}  ${v}`).join('\n');
  if (!r.liveness?.checked) return base;
  const {producing, blockBefore, blockAfter} = r.liveness;
  const livenessLine = producing
    ? pc.green('✓ chain is producing blocks') + pc.dim(` (#${blockBefore} → #${blockAfter})`)
    : pc.red(`⚠ chain is stalled at #${blockBefore} — no new blocks are being produced.`);
  return `${base}\n\n  ${livenessLine}`;
}

export async function run(opts: DoctorOptions): Promise<void> {
  const node = resolveNode(opts.node);
  const timeoutMs = opts.timeout ? Math.round(Number(opts.timeout) * 1000) : undefined;
  const checkLiveness = opts.liveness !== false;
  const t0 = Date.now();
  try {
    const report = await collectReport(node, timeoutMs, checkLiveness);
    const dt = Date.now() - t0;
    const stalled = report.liveness?.checked && !report.liveness.producing;
    if (opts.json) {
      console.log(JSON.stringify(report, null, 2));
    } else {
      console.log();
      console.log(pc.bold('  pdk-ts doctor  ') + pc.dim(`(${dt} ms)`));
      console.log();
      console.log(renderText(report));
      console.log();
    }
    if (stalled) {
      await closeApi();
      process.exit(1);
    }
  } catch (err) {
    const msg = humanizeChainError(err, node);
    if (opts.json) {
      console.log(JSON.stringify({error: msg, endpoint: node}, null, 2));
    } else {
      console.error(pc.red(`\n  ✗ doctor failed — ${msg}\n`));
      console.error(pc.dim(`  endpoint: ${node}\n`));
    }
    await closeApi();
    // WsProvider keeps sockets alive on connect failure — force clean exit
    // rather than let the event loop hang the process.
    process.exit(1);
  }
  await closeApi();
}
